domain/secret/service/watcher.go

```diff
@@ -26,23 +54,139 @@ func (s *SecretService) WatchConsumedSecretsChanges(ctx context.Context, unitNam
 // WatchRemoteConsumedSecretsChanges watches secrets remotely consumed by any unit
 // of the specified app and retuens a watcher which notifies of secret URIs
 // that have had a new revision added.
-func (s *SecretService) WatchRemoteConsumedSecretsChanges(ctx context.Context, appName string) (watcher.StringsWatcher, error) {
+func (s *WatchableService) WatchRemoteConsumedSecretsChanges(ctx context.Context, appName string) (watcher.StringsWatcher, error) {
 	ch := make(chan []string, 1)
 	ch <- []string{}
 	return watchertest.NewMockStringsWatcher(ch), nil
 }
 
 // WatchObsolete returns a watcher for notifying when:
 //   - a secret owned by the entity is deleted
-//   - a secret revision owed by the entity no longer
+//   - a secret revision owned by the entity no longer
 //     has any consumers
 //
 // Obsolete revisions results are "uri/revno" and deleted
 // secret results are "uri".
-func (s *SecretService) WatchObsolete(ctx context.Context, owners ...CharmSecretOwner) (watcher.StringsWatcher, error) {
-	ch := make(chan []string, 1)
-	ch <- []string{}
-	return watchertest.NewMockStringsWatcher(ch), nil
+func (s *WatchableService) WatchObsolete(ctx context.Context, owners ...CharmSecretOwner) (watcher.StringsWatcher, error) {
+	if len(owners) == 0 {
+		return nil, errors.New("at least one owner must be provided")
+	}
+
+	appOwners, unitOwners := splitCharmSecretOwners(owners...)
+	table, stmt := s.st.InitialWatchStatementForObsoleteRevision(ctx, appOwners, unitOwners)
+	w, err := s.watcherFactory.NewNamespaceWatcher(
+		table, changestream.Update, stmt,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	f := func(ctx context.Context, revisionUUIDs ...string) ([]string, error) {
+		return s.st.GetRevisionIDsForObsolete(ctx, appOwners, unitOwners, revisionUUIDs...)
+	}
+	return newObsoleteRevisionWatcher(w, s.logger, f)
+}
+
+type obsoleteRevisionWatcher struct {
+	catacomb catacomb.Catacomb
+	logger   Logger
+
+	sourceWatcher  watcher.StringsWatcher
+	processChanges func(
+		ctx context.Context,
+		revisionUUID ...string,
+	) ([]string, error)
+
+	out chan []string
+}
+
+func newObsoleteRevisionWatcher(
+	sourceWatcher watcher.StringsWatcher, logger Logger,
+	processChanges func(ctx context.Context, revisionUUID ...string) ([]string, error),
+) (*obsoleteRevisionWatcher, error) {
+	w := &obsoleteRevisionWatcher{
+		sourceWatcher:  sourceWatcher,
+		logger:         logger,
+		processChanges: processChanges,
+		out:            make(chan []string),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{sourceWatcher},
+	})
+	return w, errors.Trace(err)
+}
+
+func (w *obsoleteRevisionWatcher) loop() (err error) {
+	defer close(w.out)
+
+	var changes set.Strings
+	var out chan<- []string
+	ctx := w.catacomb.Context(context.Background())
```

Comment:
This should be a scoped context with cancel
```
func (w *obsoleteRevisionWatcher) scopedContext() (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(context.Background())
	return w.catacomb.Context(ctx), cancel
}
```
Then, below where ctx is used, the method needs to do
```
func (w *obsoleteRevisonWatcher) processChanges() error {
	ctx, cancel := w.scopedContext()
	defer cancel()
         ...
}
```